### ISSUE TITLE
S3SinkSpec: S3Sink should retry part upload after a transient downstream connection error

### ISSUE THREAD

**Issue by raboof** [akka/alpakka#2209]
Failed at https://travis-ci.com/github/akka/alpakka/jobs/298526637#L916

```
[info] S3Sink
[info] - should retry part upload after a transient downstream connection error *** FAILED *** (1 minute)
[info]   A timeout occurred waiting for a future to complete. Queried 101 times, sleeping 600 milliseconds between each query. (S3SinkSpec.scala:75)
[info]   org.scalatest.concurrent.Futures$FutureConcept$$anon$1:
[info]   at org.scalatest.concurrent.Futures$FutureConcept.tryTryAgain$1(Futures.scala:538)
[info]   at org.scalatest.concurrent.Futures$FutureConcept.futureValueImpl(Futures.scala:550)
[info]   at org.scalatest.concurrent.Futures$FutureConcept.futureValueImpl$(Futures.scala:479)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValueImpl(ScalaFutures.scala:275)
[info]   at org.scalatest.concurrent.Futures$FutureConcept.futureValue(Futures.scala:476)
[info]   at org.scalatest.concurrent.Futures$FutureConcept.futureValue$(Futures.scala:475)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValue(ScalaFutures.scala:275)
[info]   at docs.scaladsl.S3SinkSpec.$anonfun$new$4(S3SinkSpec.scala:75)
```

Possibly related to #2114 / #2116

**Comment by bpiper** [comment]
I've tried replicating this failure locally on latest master and the commit that the build failed on, but without success.  It doesn't appear to have failed since the build mentioned above, but it did also fail on a couple of older builds, and on at least one occasion, `retry part upload no more than the configured number of times` also failed in the same build.  Both tests are quite reliant on Wiremock doing its thing (receiving multiple requests in quick succession), so it is conceivable that there is some sort of fragility/race-condition there.

I'll try to keep an eye on this and if it fails again, perhaps I can tweak the test cases to add delay between HTTP requests or something like that.

**Comment by seglo** [comment]
I added a ScalaTest trait that can be used to repeat a test to help recreate transient test failures.  If you're interested in working on this issue then mixin the [`Repeated`](https://github.com/akka/alpakka/blob/master/testkit/src/main/scala/akka.stream.alpakka.testkit/scaladsl/Repeated.scala) trait from the testkit and run the test with `-DtimesToRepeat=n`.

**Comment by bpiper** [comment]
Thanks @seglo, that's rather useful.  No luck replicating the failure after a few thousand runs though, so might just have to wait and see if it's still failing in Travis.

**Comment by ennru** [comment]
Failed again on master https://travis-ci.com/github/akka/alpakka/jobs/324847186#L918

**Comment by ennru** [comment]
Hopefully solved by backing off in #2282

**Comment by bpiper** [comment]
Unfortunately not... despite being successful in 20+ branch builds, the test has already failed in 2 master builds, for reasons I can't quite fathom yet.  This should probably be re-opened.

**Comment by bpiper** [comment]
@ennru do you know if there's anything special/distinguishing about the master build compared to branch builds? Anything that might impact test execution at least.

I'm going to try switching to the latest version of the `wiremock-jre8` build, since the `wiremock` one is supposedly for pre Java 8 runtimes, and this issue looked somewhat relevant - https://github.com/tomakehurst/wiremock/issues/970.

If that doesn't help, then it might be necessary to disable the failing test for now.

**Comment by raboof** [comment]
> do you know if there's anything special/distinguishing about the master build compared to branch builds? Anything that might impact test execution at least.

Not AFAICS. I've cleared the caches.

**Comment by bpiper** [comment]
At the risk of jinxing things, the S3 tests haven't failed in any of the 21 master builds since the latest PR was merged, so I think this can probably be closed now (at least until #22 fails and furthers my descent into superstitious paranoia).

**Comment by raboof** [comment]
I agree, thanks!
